examples/version-migration/separate-container/src/start.ts

```diff
@@ -77,84 +132,51 @@ const renderModel = (model: IVersionedModel, migrationTool: IMigrationTool): voi
 	}
 };
 
-async function start(): Promise<void> {
-	const loader = new SimpleLoader({
-		urlResolver: new InsecureTinyliciousUrlResolver(),
-		documentServiceFactory: new RouterliciousDocumentServiceFactory(
-			new InsecureTinyliciousTokenProvider(),
-		),
-		codeLoader: new DemoCodeLoader(),
-		generateCreateNewRequest: createTinyliciousCreateNewRequest,
+export const setupContainer = async (
+	id: string,
+	alreadyLoadedContainer?: IContainer | undefined,
+): Promise<void> => {
+	// The first createDetached flow ends up with a live container reference that we want to retain rather
+	// than disposing it and loading a second time.  In all other cases we'll do the actual load here.
+	const container =
+		alreadyLoadedContainer ??
+		(await loadExistingContainer({ ...loaderProps, request: { url: id } }));
+	const model = await getModelFromContainer<IMigratableModel>(container);
+
+	// In this example, our container code mixes in an IMigratorEntryPoint to the container entryPoint.  The getMigrator
+	// function lets us construct an IMigrator by providing the necessary external tools it needs to operate.  The IMigrator
+	// is an object we can use to watch migration status, propose a migration, and discover the migration result.
+	const { getMigrator } = (await container.getEntryPoint()) as IMigratorEntryPoint;
+	const migrator: IMigrator = await getMigrator(
+		async () => loadExistingContainer({ ...loaderProps, request: { url: id } }),
+		migrationCallback,
+	);
+	migrator.events.on("migrated", () => {
+		const newContainerId = migrator.migrationResult as string;
+		container.dispose();
+		setupContainer(newContainerId).catch(console.error);
 	});
 
+	renderModel(model, migrator);
+	updateTabForId(id);
+};
+
+async function start(): Promise<void> {
 	let id: string;
-	let model: IMigratableModel;
-	let migrationTool: IMigrationTool;
+	let container: IContainer;
 
 	if (location.hash.length === 0) {
 		// Choosing to create with the "old" version for demo purposes, so we can demo the upgrade flow.
 		// Normally we would create with the most-recent version.
-		const { container, attach } = await loader.createDetached("one");
-		const modelAndMigrationTool =
-			await getModelAndMigrationToolFromContainer<IMigratableModel>(container);
-		model = modelAndMigrationTool.model;
-		migrationTool = modelAndMigrationTool.migrationTool;
-		id = await attach();
+		const createDetachedResult = await createDetachedCallback("one");
+		container = createDetachedResult.container;
+		id = await createDetachedResult.attach();
 	} else {
 		id = location.hash.slice(1);
-		const container = await loader.loadExisting(id);
-		const modelAndMigrationTool =
-			await getModelAndMigrationToolFromContainer<IMigratableModel>(container);
-		model = modelAndMigrationTool.model;
-		migrationTool = modelAndMigrationTool.migrationTool;
+		container = await loadExistingContainer({ ...loaderProps, request: { url: id } });
```

Comment:
Looks like loadExistingContainer could be skipped. setupContainer already does it.

Reply:
True, will remove in the next iteration.